packages/pymedphys_dicom/src/pymedphys_dicom/ct/extend.py

```diff
@@ -0,0 +1,180 @@
+# Copyright (C) 2019 Cancer Care Associates
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version (the "AGPL-3.0+").
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License and the additional terms for more
+# details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+# ADDITIONAL TERMS are also included as allowed by Section 7 of the GNU
+# Affero General Public License. These additional terms are Sections 1, 5,
+# 6, 7, 8, and 9 from the Apache License, Version 2.0 (the "Apache-2.0")
+# where all references to the definition "License" are instead defined to
+# mean the AGPL-3.0+.
+
+# You should have received a copy of the Apache-2.0 along with this
+# program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+
+
+import os
+from copy import deepcopy
+from glob import glob
+import datetime
+import random
+from collections import deque
+
+import pydicom
+from ..constants.uuid import PYMEDPHYS_ROOT_UID
+
+
+def extend(input_dir, output_dir, index_to_copy, number_of_slices):
+    filepaths = glob(os.path.join(input_dir, '*'))
+    dicom_datasets = load_dicom_into_deque(filepaths)
+    extend_datasets(dicom_datasets, index_to_copy, number_of_slices)
+
+    filenames = [
+        os.path.basename(filepath)
+        for filepath in filepaths
+    ]
+
+    common_prefix = os.path.commonprefix(filenames)
+
+    number_of_digits = len(str(len(dicom_datasets)))
+    new_filenames = [
+        "{}{}.dcm".format(
+            common_prefix,
+            str(dicom_dataset.InstanceNumber).zfill(number_of_digits)
+        )
+        for dicom_dataset in dicom_datasets
+    ]
+
+    new_filepaths = [
+        os.path.join(output_dir, filename)
+        for filename in new_filenames
+    ]
+
+    for dicom_dataset, filepath in zip(dicom_datasets, new_filepaths):
+        dicom_dataset.save_as(filepath)
+
+
+def extend_datasets(dicom_datasets, index_to_copy, number_of_slices, uids=None):
+    copy_slices_and_append(dicom_datasets, index_to_copy, number_of_slices)
+    refresh_instance_numbers(dicom_datasets)
+    generate_new_uids(dicom_datasets, uids=uids)
+
+
+def load_dicom_into_deque(filepaths):
+    dicom_datasets_initial_read = [
+        pydicom.dcmread(filepath, force=True)
+        for filepath in filepaths
+    ]
+
+    dicom_datasets = convert_datasets_to_deque(dicom_datasets_initial_read)
+
+    return dicom_datasets
+
+
+def convert_datasets_to_deque(datasets):
+    dicom_datasets = deque()
```

Comment:
A little confusing. `datasets` is a list but `dicom_datasets` is a `deque`.

Maybe replace `dicom_datasets` with `dataset_deque`?

Reply:
See https://stackoverflow.com/a/8793247/3912576

Reply:
All for changing the name, but I'm not keen on using hungarian typing...

Reply:
Really? We have quite a number of examples in `pymedphys` where we *have* done this. Particularly for dictionaries...

Not to mention that any time we use a variable name like `dataset` for an instance of class `Dataset`, we are basically doing the same thing as my suggestion above. We have dozens of examples of `ds_<description>`. Would you say the `ds` should be removed? Surely not...

The SO post makes sense for simple data types where type is either (clearly) implicit or you want to maintain type flexibility (e.g. you might be happy for an `int` to be cast as a `float` and so `age_int` would be stupid). But I really don't think this holds for more complex abstract data types, especially where you want to maintain the specific type (otherwise functionality will break). I would argue this is the case for your use of `deque` here...